hxunxmlns.1
.de d \" begin display
.sp
.in +4
.nf
..
.de e \" end display
.in -4
.fi
.sp
..
.TH "HXUNXMLNS" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxunxmlns \- replace XML "global names" by Namespace prefixes
.SH SYNOPSIS
.B hxunxmlns
.RB "[\| " \-\- " \|]"
.RI "[\| " file " \|]"
.SH DESCRIPTION
.B hxunxmlns
is the reverse of
.BR hxxmlns (1).
It converts element and attribute names of the form
.d
<{URL1}foo {URL2}bar="...">
.e
to
.d
<foo xmlns="URL1" xmlns:x="URL2" x:bar="...">
.e
.PP
The predefined XML Namespace is recognized and replaced by the prefix "xml:".
.SH "DIAGNOSTICS"
The following exit values are returned:
.TP 10
.B 0
Successful completion.
.TP
.B > 0
One or more errors occurred in the parsing of the file.
.SH "SEE ALSO"
.BR hxxmlns (1)
.SH BUGS
.LP
The error recovery for incorrect XML is primitive.


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxunxmlns.c
/*
 * unxmlns - convert <{namespace}foo> to <foo xmlns="namespace">
 *
 * This program is the reverse of xmlns.
 *
 * To do: optimize, i.e., reuse inherited namespace declaration,
 * instead of declaring them again on every element.
 *
 * Copyright © 2005 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Author: Bert Bos
 * Created: 8 November 2005
 * Version: $Id: hxunxmlns.c,v 1.7 2017/11/24 09:50:25 bbos Exp $
 *
 **/
#include "config.h"
#include <stdio.h>
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#include <ctype.h>
#if STDC_HEADERS
# include <string.h>
#else
# ifndef HAVE_STRCHR
#  define strchr index
#  define strrchr rindex
# endif
#endif
#include <stdlib.h>
#include <assert.h>
#include <stdbool.h>
#include "export.h"
#include "types.e"
#include "html.e"
#include "scan.e"
#include "dict.e"
#include "openurl.e"
#include "errexit.e"


#define XMLNS "{http://www.w3.org/XML/1998/namespace}"
const static size_t XMLNSLEN = 39;		/* strlen(XMLNS) */

extern int yylineno;				/* From scan.l */

static bool has_error = false;


/* --------------- implements interface api.h -------------------------- */

/* handle_error -- called when a parse error occurred */
void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
  has_error = true;
}

/* start -- called before the first event is reported */
void* start(void)
{
  return NULL;
}
  
/* end -- called after the last event is reported */
void end(void *clientdata)
{
  /* skip */
}

/* handle_comment -- called after a comment is parsed */
void handle_comment(void *clientdata, string commenttext)
{
  printf("<!--%s-->", commenttext);
}

/* handle_text -- called after a text chunk is parsed */
void handle_text(void *clientdata, string text)
{
  fputs(text, stdout);
}

/* handle_decl -- called after a declaration is parsed */
void handle_decl(void *clientdata, string gi, string fpi,
		 string url)
{
  printf("<!DOCTYPE %s", gi);
  if (fpi) printf(" PUBLIC \"%s\"\n", fpi);
  if (url) printf(" %s\"%s\"", fpi ? "" : "SYSTEM ", url);
  printf(">");
}

/* handle_pi -- called after a PI is parsed */
void handle_pi(void *clientdata, string pi_text)
{
  printf("<?%s>", pi_text);
}

/* print_attrs -- print attributes and declare their namespaces, if any */
static void print_attrs(const pairlist attribs)
{
  string h, s;
  pairlist p;
  int n = 0;

  for (p = attribs; p; p = p->next) {
    if (p->name[0] != '{') {
      printf(" %s=\"%s\"", p->name, p->value);
    } else if (p->name[1] == '}') {
      printf(" %s=\"%s\"", p->name + 2, p->value);
    } else if (strncmp(p->name, XMLNS, XMLNSLEN) == 0) {
      printf(" xml:%s=\"%s\"", p->name + XMLNSLEN, p->value);
    } else if (! (h = strchr(p->name, '}'))) {
      fprintf(stderr, "%d: Unmatched \"{\" in attribute name (\"%s\")\n",
	      yylineno, p->name);
      has_error = true;
    } else {
      printf(" xmlns:x%d=\"", n);
      for (s = p->name + 1; s != h; s++) putchar(*s);
      printf("\" x%d:%s=\"%s\"", n, h + 1, p->value);
    }
  }
}

/* print_tag_start -- print "<foo" and possibly an xmlns="..." */
static void print_tag_start(const string name)
{
  string s, h;

  if (name[0] != '{') {
    printf("<%s", name);
  } else if (name[1] == '}') {
    printf("<%s", name + 2);
  } else if (strncmp(name, XMLNS, XMLNSLEN) == 0) {
    printf("<xml:%s", name + XMLNSLEN);
  } else if (! (h = strchr(name, '}'))) {
    fprintf(stderr, "%d: Unmatched \"{\" in tag name (\"%s\")\n",
	    yylineno, name);
    has_error = true;
  } else {
    printf("<%s xmlns=\"", h + 1);
    for (s = name + 1; s != h; s++) putchar(*s);
    putchar('\"');
  }
}

/* handle_starttag -- called after a start tag is parsed */
void handle_starttag(void *clientdata, string name, pairlist attribs)
{
  print_tag_start(name);
  print_attrs(attribs);
  putchar('>');
}

/* handle_emptytag -- called after an empty tag is parsed */
void handle_emptytag(void *clientdata, string name, pairlist attribs)
{
  print_tag_start(name);
  print_attrs(attribs);
  printf(" />");
}

/* handle_endtag -- called after an endtag is parsed (name may be "") */
void handle_endtag(void *clientdata, string name)
{
  string h;

  if (name[0] != '{') {
    printf("</%s>", name);
  } else if (! (h = strchr(name, '}'))) {
    fprintf(stderr, "%d: Unmatched \"{\" in tag name (\"%s\")\n",
	    yylineno, name);
    has_error = true;
  } else {
    printf("</%s>", h + 1);
  }
}

/* --------------------------------------------------------------------- */

/* usage -- print usage message and exit */
static void usage(string prog)
{
  fprintf(stderr, "Version %s\nUsage: %s [html-file-or-url]\n", VERSION, prog);
  exit(2);
}

int main(int argc, char *argv[])
{
  int i, status = 200;

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_start_handler(start);
  set_end_handler(end);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);

  /* Parse command line arguments */
  for (i = 1; i < argc && argv[i][0] == '-' && !eq(argv[i], "--"); i++) {
    switch (argv[i][1]) {
      default: usage(argv[0]);
    }
  }
  if (i < argc && eq(argv[i], "--")) i++;

  if (i == argc) yyin = stdin;
  else if (i == argc - 1 && eq(argv[i], "-")) yyin = stdin;
  else if (i == argc - 1) yyin = fopenurl(argv[i], "r", &status);
  else usage(argv[0]);

  if (yyin == NULL) {perror(argv[i]); exit(1);}
  if (status != 200) errexit("%s : %s\n", argv[i], http_strerror(status));

  if (yyparse() != 0) exit(3);

  return has_error ? 1 : 0;
}
